runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -325,74 +357,63 @@ private FlinkAutoBalancedShardKeyShardingFunction(
+        cache =
+            CacheBuilder.newBuilder()
+                .maximumSize(CACHE_MAX_SIZE)
+                .expireAfterAccess(CACHE_EXPIRE_SECONDS, TimeUnit.SECONDS)
+                .build();
+      }
+
       // we need to ensure that keys are always stable no matter at which worker they
       // are created and what is an order of observed shard numbers
-      if (cache.size() < shardNumber) {
-        for (int i = 0; i < shardNumber; i++) {
-          generateInternal(new CacheKey(destinationKey, i));
-        }
+      if (cache.getIfPresent(destinationKey) == null) {
+        cache.put(destinationKey, generateShardedKeys(destinationKey, shardCount));
       }
 
-      return generateInternal(new CacheKey(destinationKey, shardNumber));
+      return cache.getIfPresent(destinationKey).get(shardNumber);
     }
 
-    private ShardedKey<Integer> generateInternal(CacheKey key) {
-
-      ShardedKey<Integer> result = cache.get(key);
-      if (result != null) {
-        return result;
-      }
-
-      int salt = -1;
-      while (true) {
-        salt++;
-        ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key.key, salt), key.shard);
-        int targetPartition = key.shard % parallelism;
-
-        // create effective key in the same way Beam/Flink will do so we can see if it gets
-        // allocated to the partition we want
-        ByteBuffer effectiveKey;
-        try {
-          byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
-          effectiveKey = ByteBuffer.wrap(bytes);
-        } catch (CoderException e) {
-          throw new RuntimeException(e);
-        }
-
-        int partition =
-            KeyGroupRangeAssignment.assignKeyToParallelOperator(
-                effectiveKey, maxParallelism, parallelism);
-
-        if (partition == targetPartition && !usedSalts.contains(salt)) {
-          usedSalts.add(salt);
-          cache.put(key, shk);
-          return shk;
+    private Map<Integer, ShardedKey<Integer>> generateShardedKeys(int key, int shardCount) {
+
+      Set<Integer> usedSalts = new HashSet<>();
+      Map<Integer, ShardedKey<Integer>> shardedKeys = new HashMap<>();
+
+      for (int shard = 0; shard < shardCount; shard++) {
+
+        int salt = -1;
+        while (true) {
+          if (salt++ == Integer.MAX_VALUE) {
+            throw new RuntimeException(
+                "Failed to find sharded key in [ " + Integer.MAX_VALUE + " ] iterations");
+          }
+          ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key, salt), shard);
+          int targetPartition = shard % parallelism;
+
+          // create effective key in the same way Beam/Flink will do so we can see if it gets
+          // allocated to the partition we want
+          ByteBuffer effectiveKey;
+          try {
+            byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
+            effectiveKey = ByteBuffer.wrap(bytes);
```

Comment:
Could you use `FlinkKeyUtils.encodeKey(K key, Coder<K> keyCoder)` here? Just to ensure that this work correctly in case we change the encoding.

Reply:
Sorry, I forgot about this. Will do

Reply:
`FlinkKeyUtils` is package private. I assume it is OK to make it public?